crunch-core/src/main/java/org/apache/crunch/util/CrunchRenameCopyListing.java

```diff
@@ -0,0 +1,272 @@
+
+            if (isDirectoryAndNotEmpty(sourceFS, sourceStatus)) {
+              if (LOG.isDebugEnabled()) {
+                LOG.debug("Traversing non-empty source dir: " + sourceStatus.getPath());
+              }
+              traverseNonEmptyDirectory(fileListWriter, sourceStatus, sourcePathRoot, options);
+            }
+          }
+        }
+      }
+      fileListWriter.close();
+      fileListWriter = null;
+    } finally {
+      if (fileListWriter != null) {
+        try {
+          fileListWriter.close();
+        } catch(IOException e) {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("Exception in closing " + fileListWriter, e);
+          }
+        }
+      }
+    }
+  }
+
+  private Path computeSourceRootPath(FileStatus sourceStatus, DistCpOptions options) throws IOException {
+    Path target = options.getTargetPath();
+    FileSystem targetFS = target.getFileSystem(getConf());
+    final boolean targetPathExists = options.getTargetPathExists();
+
+    boolean solitaryFile = options.getSourcePaths().size() == 1 && !sourceStatus.isDirectory();
+
+    if (solitaryFile) {
+      if (targetFS.isFile(target) || !targetPathExists) {
+        return sourceStatus.getPath();
+      } else {
+        return sourceStatus.getPath().getParent();
+      }
+    } else {
+      boolean specialHandling =
+          (options.getSourcePaths().size() == 1 && !targetPathExists) || options.shouldSyncFolder() || options.shouldOverwrite();
+
+      return specialHandling && sourceStatus.isDirectory() ? sourceStatus.getPath() : sourceStatus.getPath().getParent();
+    }
+  }
+
+  private Path makeQualified(Path path) throws IOException {
+    final FileSystem fs = path.getFileSystem(getConf());
+    return path.makeQualified(fs.getUri(), fs.getWorkingDirectory());
+  }
+
+  private static boolean isDirectoryAndNotEmpty(FileSystem fileSystem, FileStatus fileStatus) throws IOException {
+    return fileStatus.isDirectory() && getChildren(fileSystem, fileStatus).length > 0;
+  }
+
+  private static FileStatus[] getChildren(FileSystem fileSystem, FileStatus parent) throws IOException {
+    return fileSystem.listStatus(parent.getPath());
+  }
+
+  private void traverseNonEmptyDirectory(SequenceFile.Writer fileListWriter, FileStatus sourceStatus, Path sourcePathRoot,
+      DistCpOptions options) throws IOException {
+    FileSystem sourceFS = sourcePathRoot.getFileSystem(getConf());
+    final boolean preserveAcls = options.shouldPreserve(FileAttribute.ACL);
+    final boolean preserveXAttrs = options.shouldPreserve(FileAttribute.XATTR);
+    final boolean preserveRawXattrs = options.shouldPreserveRawXattrs();
+    Stack<FileStatus> pathStack = new Stack<>();
+    pathStack.push(sourceStatus);
+
+    while (!pathStack.isEmpty()) {
+      for (FileStatus child : getChildren(sourceFS, pathStack.pop())) {
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("Recording source-path: " + sourceStatus.getPath() + " for copy.");
```

Comment:
should be able to do string interpolation right instead of concatenation?

Reply:
Yes, done: https://github.com/apache/crunch/pull/20/commits/643cc074597022154c7c15ac7f1f2681981d5e9d